### PR TITLE
Adding beacon implementation to encryptor hybrid

### PR DESCRIPTION
**Pull request by JP-Globality**
This PR gives us the ability to perform exact match searches on encrypted fields. It's done through the use of beacons.

Beacons are hash that's deterministically calculated such that every input has only one output.

More about beacons can be found on the aws encryption sdk docs: https://docs.aws.amazon.com/database-encryption-sdk/latest/devguide/using-beacons.html

Now we use the `*_beacon` value inside the SQL used in the query against the database e.g:
```
SELECT test_encryption_employee.id AS test_encryption_employee_id, test_encryption_employee.name_encrypted AS test_encryption_employee_name_encrypted, test_encryption_employee.name AS test_encryption_employee_name, test_encryption_employee.name_beacon AS test_encryption_employee_name_beacon
FROM test_encryption_employee
WHERE test_encryption_employee.name_beacon = %(name_beacon_1)s ORDER BY test_encryption_employee.id ASC
```

There is also support for `._in` queries using beacons e.g
```
select(Employee).filter(Employee.name.in_(["foo", "bar"]))
```
Under the hood we will beaconise each of the values and perform the search.

### CODE REVIEW

File: microcosm_postgres/encryption/v2/column.py [hunk 1]
```diff
@@ -20,6 +21,51 @@
 NOT_SET = object()
 
 
+class BeaconComparator(Comparator):
+    def __init__(
+        self,
+        val,
+        encoder_fn: Callable | None = None,
+        beacon_fn: Callable | None = None,
```
Comment (dumpstate) [inline microcosm_postgres/encryption/v2/column.py:29]: `Callable[[str], str]`

File: microcosm_postgres/encryption/v2/column.py [hunk 2]
```diff
@@ -121,3 +181,7 @@ def unencrypted(self, **kwargs: Any) -> Mapped[T | None]:
             ),
             **kwargs,
         )
+
+    def beacon(self, **kwargs: Any):
+        # TODO - assume all encrypted fields need to use beacons
```
Comment (dumpstate) [inline microcosm_postgres/encryption/v2/column.py:186]: do we still need this TODO? not sure that's a valid assumption.

**Review by dumpstate — APPROVED** [review]

File: microcosm_postgres/encryption/registry.py [hunk 3]
```diff
@@ -17,26 +17,35 @@
...
 
 
-def parse_config(context_keys: Sequence[str],
-                 key_ids: Sequence[Union[str, Sequence[str]]],
-                 account_ids: Sequence[Union[str, Sequence[str]]],
-                 partitions: Sequence[Union[str, Sequence[str]]]) -> Mapping[str, Mapping[str, Sequence[str]]]:
-    return {
-        # NB: split key id on non-comma to avoid confusion with config parsing
-        context_key: {
+def parse_config(
```
Comment (german-glob) [inline microcosm_postgres/encryption/registry.py:20]: Just a question, why we don't store the config in the format returned by this function?

File: microcosm_postgres/encryption/encryptor.py [hunk 4]
```diff
@@ -66,6 +69,14 @@ def decrypt(self, encryption_context_key: str, ciphertext: bytes) -> str:
         )
         return plaintext.decode("utf-8")
 
+    def beacon(self, value: str) -> str | None:
+        if self._beacon_key is None:
+            return None
```
Comment (german-glob) [inline microcosm_postgres/encryption/encryptor.py:74]: Is there a unit test for this case?

File: microcosm_postgres/encryption/v2/encryptors.py [hunk 5]
```diff
@@ -131,3 +138,14 @@ def decrypt(self, value: bytes) -> str:
 
         context, encryptor = self.encryptor_context
         return encryptor.decrypt(context, value)
+
+    def beacon(self, value: str) -> str:
+        if self.encryptor_context is None:
+            raise self.EncryptorNotBound()
```
Comment (german-glob) [inline microcosm_postgres/encryption/v2/encryptors.py:144]: I don't see a unit test for this error?

File: microcosm_postgres/encryption/v2/encryptors.py [hunk 6]
```diff
@@ -131,3 +138,14 @@ def decrypt(self, value: bytes) -> str:
...
+
+        _, encryptor = self.encryptor_context
+        beacon = encryptor.beacon(value)
+        if beacon is None:
+            raise self.BeaconKeyNotSet()
```
Comment (german-glob) [inline microcosm_postgres/encryption/v2/encryptors.py:149]: Is there a unit test?

File: microcosm_postgres/encryption/v2/column.py (hunk 2 above)
Comment (JP-Globality) [reply microcosm_postgres/encryption/v2/column.py:186]: nope - will strip. 👍

File: .bumpversion.cfg [hunk 7]
```diff
@@ -1,5 +1,5 @@
 [bumpversion]
-current_version = 3.5.5
+current_version = 3.6.5
```
Comment (jamie-chang-globality) [inline .bumpversion.cfg:2]: 3.6.0

File: microcosm_postgres/encryption/v2/column.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (JP-Globality) [reply microcosm_postgres/encryption/v2/column.py]: Removed - thanks

File: microcosm_postgres/encryption/v2/column.py [hunk 8]
```diff
@@ -20,6 +21,51 @@
...
+    ):
+        # Note that we only store beacon val when we are instantiating as part of
+        # the sqlalchemy model setup
+        self.beacon_val = beacon_val
+
+        self.encoder_fn = encoder_fn
+        self.beacon_fn = beacon_fn
+        self.val = val
+
+    def operate(self, op, other: Any = NOT_SET, **kwargs: Any) -> ColumnElement[Any]:  # type: ignore[override]  # noqa: E501
```
[the remark this answers is not in the document]
Comment (JP-Globality) [reply microcosm_postgres/encryption/v2/column.py:41]: it's a sqlalchemy op e.g
```
@comparison_op
@_operator_fn
def in_op(a: Any, b: Any) -> Any:
    return a.in_(b)
```

File: microcosm_postgres/encryption/v2/column.py [hunk 9]
```diff
@@ -71,19 +124,32 @@ def _prop(self):
...
+                if hasattr(self, beacon_field):
+                    setattr(self, beacon_field, None)
                 return
 
             setattr(self, encrypted_field, encrypted)
             setattr(self, unencrypted_field, None)
+            if hasattr(self, beacon_field):
+                setattr(self, beacon_field, encryptor.beacon(encoded))
+
+        def _prop_comparator(cls) -> BeaconComparator | None:
```
Comment (jamie-chang-globality) [inline microcosm_postgres/encryption/v2/column.py:141]: What happens when we return None here as the comparator?

File: microcosm_postgres/encryption/v2/column.py [hunk 10]
```diff
@@ -71,19 +124,32 @@ def _prop(self):
...
+            if beacon := getattr(cls, beacon_field, None):
+                return BeaconComparator(
+                    val=getattr(cls, unencrypted_field),
+                    beacon_val=beacon,
+                    encoder_fn=encoder.encode,
+                    beacon_fn=encryptor.beacon,
+                )
 
-        def _prop_expression(cls):
```
Comment (jamie-chang-globality) [inline microcosm_postgres/encryption/v2/column.py:83]: Should we keep the expression for when we don't use beacon?

File: microcosm_postgres/encryption/v2/column.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (JP-Globality) [reply microcosm_postgres/encryption/v2/column.py:41]: Added `Callable` type to cover. Not going to attempt covering all different the flavours of op 😅

File: microcosm_postgres/encryption/registry.py (hunk 3 above)
Comment (JP-Globality) [reply microcosm_postgres/encryption/registry.py:20]: This is some legacy stuff left over from the old encryption v1 days. We create a new structure that keys by the client id so it's easier to use later on.

File: microcosm_postgres/tests/test_employee_store_with_encryption.py [hunk 11]
```diff
@@ -0,0 +1,276 @@
...
+    with AwsKmsEncryptor.set_encryptor_context("test", single_tenant_encryptor):
+        session.add(Employee(name="foo", salary=1000))
+        session.add(Employee(name="bar", salary=1000))
+        session.commit()
+
+    query = select(Employee).filter(Employee.salary == 1000).order_by(Employee.name.asc())
+    results = session.execute(query).scalars().all()
+
+    assert len(results) == 0
+
```
Comment (jamie-chang-globality) [inline microcosm_postgres/tests/test_employee_store_with_encryption.py:251]: I think you need a test where we add employee(s) without encryption and then we do the search

I'm not sure it works in this case?

File: microcosm_postgres/encryption/v2/column.py [hunk 12]
```diff
@@ -121,3 +187,6 @@ def unencrypted(self, **kwargs: Any) -> Mapped[T | None]:
             ),
             **kwargs,
         )
+
+    def beacon(self, **kwargs: Any):
```
Comment (jamie-chang-globality) [inline microcosm_postgres/encryption/v2/column.py:191]: typing:
` -> Mapped[str | None]:`

For release 2 just ` -> Mapped`.

File: microcosm_postgres/encryption/v2/column.py [hunk 13]
```diff
@@ -20,6 +21,51 @@
...
+        self.encoder_fn = encoder_fn
+        self.beacon_fn = beacon_fn
+        self.val = val
+
+    def operate(self, op, other: Any = NOT_SET, **kwargs: Any) -> ColumnElement[Any]:  # type: ignore[override]  # noqa: E501
+        # This first condition might happen when we are doing an order by ACS / DESC
+        if other is NOT_SET:
+            return op(self.beacon_val)
+        if not isinstance(other, BeaconComparator):
+            other = BeaconComparator(other)
```
[the remark this answers is not in the document]
Comment (JP-Globality) [reply microcosm_postgres/encryption/v2/column.py]: correct each case should be covered.

Covered by the order by
```
        if other is NOT_SET:
            return op(self.beacon_val)
```

Covered by the `.in` unit test
```
return op(self.beacon_val, [self._beaconise(v) for v in other], **kwargs)
```

File: microcosm_postgres/encryption/encryptor.py (hunk 4 above)
Comment (JP-Globality) [reply microcosm_postgres/encryption/encryptor.py:74]: Not explicitly, let me create one.

File: microcosm_postgres/encryption/v2/encryptors.py (hunk 5 above)
Comment (JP-Globality) [reply microcosm_postgres/encryption/v2/encryptors.py:144]: Have added one. thanks

File: microcosm_postgres/encryption/v2/encryptors.py (hunk 6 above)
Comment (JP-Globality) [reply microcosm_postgres/encryption/v2/encryptors.py:149]: added explicit unit test for this. Thanks

File: microcosm_postgres/encryption/v2/column.py (hunk 10 above)
Comment (JP-Globality) [reply microcosm_postgres/encryption/v2/column.py:83]: We only know whether to use beacon lower down in the call stack e.g in the comparator on an object instance?

I've put more logic in the comparator to deal with using / not using it? I would be simpilar to just to use prop expression though if you know a way?

File: microcosm_postgres/tests/test_employee_store_with_encryption.py (hunk 11 above)
Comment (JP-Globality) [reply microcosm_postgres/tests/test_employee_store_with_encryption.py:331]: True, it didn't work. I added a test case to cover and updated the implementation.

File: microcosm_postgres/encryption/v2/column.py (hunk 9 above)
Comment (JP-Globality) [reply microcosm_postgres/encryption/v2/column.py]: I've updated this so we don't return None. Type was incorrect.

File: setup.py [hunk 14]
```diff
@@ -33,7 +33,10 @@
     ],
     extras_require={
         "metrics": "microcosm-metrics>=2.5.0",
-        "encryption": "aws-encryption-sdk>=2.0.0",
+        "encryption": [
+            "aws-encryption-sdk>=2.0.0",
+            "cryptography>=41",
```
Comment (vnguyen-globality) [inline setup.py:38]: I had to bump `pyOpenSSL` to 23.2.0 due to dependency mismatch ( with 22~ it required cryptopgraphy to be < 40 for me)

**Comment by vnguyen-globality** [comment]
I'm encountering a problem with `Store`s that have an auto_filter_fields on a column that has a beacon. 

https://github.com/globality-corp/microcosm-postgres/blob/e3ec0ab5b846f703807062d66e14a9a337d7a1a9/microcosm_postgres/store.py#L41C8-L41C8

AttributeError("Neither 'encryption' object nor 'BeaconComparator' object associated with [some_model_class] has an attribute 'name'")

The hybrid property has the comparator at the top level instead? 

`(Pdb) pp auto_filter_fields
(<sqlalchemy.orm.attributes.InstrumentedAttribute object at 0x107a3eb60>,
 <sqlalchemy.orm.attributes.InstrumentedAttribute object at 0x107a3eca0>,
 <sqlalchemy.orm.attributes.create_proxied_attribute.<locals>.Proxy object at 0x10924aae0>,
 <sqlalchemy.orm.attributes.InstrumentedAttribute object at 0x107a3ec00>,
 <sqlalchemy.orm.attributes.InstrumentedAttribute object at 0x107a3ed40>,
 <sqlalchemy.orm.attributes.InstrumentedAttribute object at 0x107a3f2e0>)

(Pdb) pp vars(auto_filter_fields[2])
{'__doc__': None,
 '_comparator': <microcosm_postgres.encryption.v2.column.BeaconComparator object at 0x109255170>,
 'comparator': <microcosm_postgres.encryption.v2.column.BeaconComparator object at 0x109255170>,
 'descriptor': <microcosm_postgres.encryption.v2.column.encryption object at 0x107ad4490>,
 'original_property': None}`
 
 
 without beacon, the comparator ends up being the regular column instead? 
 `
 (Pdb) pp auto_filter_fields
(<sqlalchemy.orm.attributes.InstrumentedAttribute object at 0x1085d6c00>,
 <sqlalchemy.orm.attributes.InstrumentedAttribute object at 0x1085d6d40>,
 <sqlalchemy.orm.attributes.create_proxied_attribute.<locals>.Proxy object at 0x109dd2a30>,
 <sqlalchemy.orm.attributes.InstrumentedAttribute object at 0x1085d6ca0>,
 <sqlalchemy.orm.attributes.InstrumentedAttribute object at 0x1085d6de0>,
 <sqlalchemy.orm.attributes.InstrumentedAttribute object at 0x1085d72e0>)

(Pdb) pp vars(auto_filter_fields[2])
{'__doc__': None,
 '_comparator': <sqlalchemy.orm.attributes.InstrumentedAttribute object at 0x1085d6e80>,
 'comparator': <sqlalchemy.orm.attributes.InstrumentedAttribute object at 0x1085d6e80>,
 'descriptor': <microcosm_postgres.encryption.v2.column.encryption object at 0x108668490>,
 'original_property': None}


{'_Annotated__element': Column('purchase_order_id', String(), table=<external_purchase_order>),
 '_annotations': immutabledict({'entity_namespace': <Mapper at 0x10866b640; ExternalPurchaseOrder>, 'parententity': <Mapper at 0x10866b640; ExternalPurchaseOrder>, 'parentmapper': <Mapper at 0x10866b640; ExternalPurchaseOrder>, 'proxy_key': 'purchase_order_id_unencrypted'}),
 ... 
 'key': 'purchase_order_id',
 'name': 'purchase_order_id',
 'nullable': True,
 ... 
 `

Not sure how to remedy this. Maybe in the `Store` during the init we somehow pull out `comparator.beacon_val` ?

**Pull request merged**